### DISCUSSION TITLE
LD240

### DISCUSSION THREAD

**Discussion by BobbyBobJohn** [Nortank12/deepcool-digital-linux#9] [Ideas]
Hi,

LD240 works well, but had issues with running on start/wake. Made it run as non-root for everything to work properly. Is it possible to add a flag to startup arguments to bypass root check for someone who may need it at some point?

**Comment by Nortank12** [comment]
Hi,
Yes, it would be possible to add a flag however, an even better solution would be to run it without ever requiring root access.
The problem comes from accessing the device. I looked into udev rules but haven't found a "universal" solution to bypass root. Although it could be scripted, that solution doesn't seem satisfying.

How did you manage to access your cooler as user?

**Reply by BobbyBobJohn** [reply]
First need write access to the display. It appears as HID device. So I just added a rule in udev.

/etc/udev/rules.d/99-hidraw-permissions.rules

KERNEL=="hidraw*", SUBSYSTEM=="hidraw", MODE="0660", GROUP="plugdev"

Obviously, you can just put VID:PID values in there, but I couldn't be bothered.

Then to read the power state:


    sudo apt install sysfsutils
    Add this line in /etc/sysfs.conf : mode class/powercap/intel-rapl:0/energy_uj = 0444
    reboot

I pulled the source and rebuilt without root check. That's pretty much it.

Just want to thank you for sharing your utility. Without it I doubt I would have purchased this cooler.

**Reply by Nortank12** [reply]
Thank you for the kind words!

I spent some time to learn about udev rules and managed to find a solution. I changed the udev rule to be the following:
```bash
# Intel RAPL energy usage file
ACTION=="add", SUBSYSTEM=="powercap", KERNEL=="intel-rapl:0", RUN+="/bin/chmod 444 /sys/class/powercap/intel-rapl/intel-rapl:0/energy_uj"

# DeepCool HID raw devices
SUBSYSTEM=="hidraw", ATTRS{idVendor}=="3633", MODE="0666"
```

This should cover all the issues when running it rootless, however rules are located in `/lib/udev/rules.d` on both Arch and Debian (the systems that I use). Which distribution do you use?

**Reply by BobbyBobJohn** [reply]
Ubuntu 22.04. The udev rules are in /etc/udev/rules.d

**Reply by Nortank12** [reply]
Thank you. I decided to put some instructions in the README and instead of forcing root access, I place a warning if the user doesn't have permission to read/write the device.
One more thing: did you put the service inside the `/etc/systemd/system/` folder or it runs later?

**Reply by BobbyBobJohn** [reply]
I tried a bunch of stuff to get it to work with "depends on", waiting for file, etc. For some reason, the HID device always comes up after so when your program runs it says "device not found". I think that's how I got to this stage of running it as user. I gave up on systemd config. I then used autostart in Plasma and system-sleep for suspend/resume.

/lib/systemd/system-sleep/

It's working pretty well now - haven't had to touch it at all

**Reply by Nortank12** [reply]
I removed the code that checks the privilege, so it will work without root. I've been limited on time recently, but have a few updates in mind. One of them is updating how the error messages look. For now I created a release so it can be used.

Thanks for your help!

**Reply by BobbyBobJohn** [reply]
Awesome! I'm sure it will make the setup much easier.
On my side, your program gets killed when the machine goes to sleep and I have to restart it in systemd. You may want to add that in case someone encounters the same issue.

**Reply by Nortank12** [reply]
I will test it as well when I have more time and will try to reproduce the issue.

**Reply by Nortank12** [reply]
I managed to test Ubuntu 22.04 and 24.04 and neither of them had an issue with start/wake. I used a fresh install and copy-pasted the steps from the repository (without rootless mode) and tried to reboot and suspend the PC multiple times.

Your issue might be unique due to configuration differences or interfering apps. For that reason I will not add this to the instructions but if it turns out to be a bigger problem affecting more systems and I can reproduce the issue, I will create a section for it.

Thanks for your help once again!

**Reply by BobbyBobJohn** [reply]
Ok, great.
Interestingly, I had installed Ubuntu Studio LTS (I think it's 24.04) and I had the same issue where I had to put it into the wake script.
But I'm glad it's all working and I must thank you for the utility once again.
